### PR TITLE
[WIP] Unit Test for dashboard

### LINKED ISSUE #667: [Dashboard] Missing Unit Testing
Can add with selenium library, as mentioned here: https://github.com/Kitware/trame/discussions/411

Create tests for:
- A CSR example
- A space charge example
- FODO example
- Loading files
- Dashboard stess test?

### PR DESCRIPTION
**Pull request by proy30**
This PR introduces the first unit test for the ImpactX dashboard, leveraging the SeleniumBase library to simulate a simulation run by a user with the GUI.

### Key Features
  - The unit test runs a simulation using the parameters from the FODO cell example on the GUI. (takes ~1 minute)
  - The test is considered successful if the simulation is completed without errors.

### Issues

- ~~**Dependency on Running Dashboard:**~~
  - ~~The current implementation requires the ImpactX dashboard to run in a separate terminal before executing the test via `pytest`.~~
  - ~~Can investigate to potentially have dashboard be launched within the test itself.~~
  - ~~Unit test is flaky with passing/failing, unsure if this is due to code error or error in pytest implementation. Need to investigate. Started occurring after the implementation of subprocess.~~
 - **Adding seleniumbase as a dependency**
   - Unsure where to add. (implemented incorrectly)

### Checklist

- [X] Implement unit test for impactx-dashboard (testing with FODO cell simulation parameters.
- [X] Automate the dashboard launch within the test.
- [X] Fix flaky unit test
- [x] Include seleniumbase dependency in CI test files
- [x] Include trame dependencies in CI test files

Resolves #667

### CODE REVIEW

File: tests/python/dashboard/util.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
+def wait_for_ready(sb, timeout=60):
+    for i in range(timeout):
+        print(f"wait_for_ready {i}")
+        if sb.is_element_present(".trame__loader"):
+            sb.sleep(1)
+        else:
+            print("Ready")
```
Comment (proy30) [inline tests/python/dashboard/util.py:21]: This is code directly from https://github.com/Kitware/trame-client/blob/master/trame_client/utils/testing.py. Not sure of the process to credit, if any.

File: .github/workflows/stubs.yml [hunk 2]
```diff
@@ -41,6 +41,10 @@ jobs:
         python3 -m pip install -U pip setuptools wheel
         python3 -m pip install -U pip mpi4py pytest pybind11-stubgen pre-commit seleniumbase
 
+    - name: Dashboard Dependencies
+      run: |
+        python3 -m pip install -U trame trame-matplotlib trame-plotly trame-router trame-xterm trame-vuetify
```
Comment (proy30) [inline .github/workflows/stubs.yml]: It seems that adding this caused stub files test to continuously run.

File: tests/python/dashboard/util.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    sb.update_text(selector, value)
+    sb.send_keys(selector, "\n")
+
+def start_dashboard():
+    """
+    Function which starts up impactx-dashboard server.
+    """
+    script_dir = os.path.dirname(os.path.abspath(__file__))
+    working_directory = os.path.join(script_dir, "../../../src/python/impactx")
+    working_directory = os.path.normpath(working_directory)
```
Comment (ax3l) [inline tests/python/dashboard/util.py:80]: We need to generalize this so it:
- runs in CTests from the build directory
- runs when copied in a package manager with external installs

File: tests/python/dashboard/test_dashboard.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+import time
+
+def test_simulation():
+    """
+    This test runs the FODO example on the dashboard and verifies
+    that the simulation has ran successfully.
+    """
+
+    app_process = start_dashboard()
+    time.sleep(10)
```
Comment (ax3l) [inline tests/python/dashboard/test_dashboard.py]: This is very long.

File: tests/python/dashboard/util.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+    return subprocess.Popen(
+        ["python", "-m", "dashboard"],
```
Comment (ax3l) [inline tests/python/dashboard/util.py]: When I run this locally, this literally opens a browser window for me.
Is there a way to run this headless?

File: tests/python/dashboard/util.py (hunk 1 above)
Comment (ax3l) [reply tests/python/dashboard/util.py:21]: Add inline comments with links? :)

File: tests/python/dashboard/test_dashboard.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
+from seleniumbase import SB
+from util import set_input_value, start_dashboard, wait_for_ready
+import time
+
+def test_simulation():
```
Comment (ax3l) [inline tests/python/dashboard/test_dashboard.py]:
```suggestion
def test_dashboard():

```

File: .github/workflows/stubs.yml [hunk 7]
```diff
@@ -38,8 +38,7 @@ jobs:
     - name: Dependencies
       run: |
         .github/workflows/dependencies/gcc-openmpi.sh
-        python3 -m pip install -U pip setuptools wheel
-        python3 -m pip install -U pip mpi4py pytest pybind11-stubgen pre-commit
+        python3 -m pip install -U pybind11-stubgen
```
Comment (ax3l) [inline .github/workflows/stubs.yml]:
```suggestion
        python3 -m pip install -U pybind11-stubgen pre-commit
```

Comment (ax3l) [reply .github/workflows/stubs.yml]: done

File: .github/workflows/stubs.yml (hunk 2 above)
Comment (ax3l) [reply .github/workflows/stubs.yml]: fixed

File: tests/python/dashboard/test_dashboard.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+        # Run simulation
+        sb.click("#Run_route")
+        sb.sleep(1)
+        sb.click("#run_simulation_button")
+        sb.sleep(7)
+
+        # Check if "Simulation complete" message is printed
+        xterm_content = sb.get_text("#xterm_component")
+        assert "Simulation complete." in xterm_content
```
Comment (proy30) [inline tests/python/dashboard/test_dashboard.py]: This is currently buggy and offering flaky results.

File: tests/python/dashboard/test_dashboard.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    This test runs the FODO example on the dashboard and verifies
+    that the simulation has ran successfully.
+    """
+    from seleniumbase import SB
+
+    app_process = start_dashboard()
+    time.sleep(10)
+
+    try:
+        with SB() as sb:
```
Comment (ax3l) [inline tests/python/dashboard/test_dashboard.py]: Trying if this prevents that the unit test on Linux/macOS opens a browser:
```suggestion
        with SB(headless=True) as sb:

```

File: tests/python/dashboard/test_dashboard.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    importlib.util.find_spec("seleniumbase") is None, reason="seleniumbase is not available"
+)
+def test_simulation():
+    """
+    This test runs the FODO example on the dashboard and verifies
+    that the simulation has ran successfully.
+    """
+    from seleniumbase import SB
+
+    app_process = start_dashboard()
```
Comment (ax3l) [inline tests/python/dashboard/test_dashboard.py]: Should this line go down into the `with SB() as sb:` block?

File: tests/python/dashboard/test_dashboard.py (hunk 9 above)
Comment (ax3l) [reply tests/python/dashboard/test_dashboard.py]: https://github.com/ECP-WarpX/impactx/pull/672/files#r1717652047

File: tests/python/dashboard/test_dashboard.py (hunk 8 above)
Comment (proy30) [reply tests/python/dashboard/test_dashboard.py]: Alternative method implemented by having unit test interact with phase space projections, which indirectly mean simulation ran properly.

**Comment by ax3l** [comment]
Continued in #706

**Pull request closed without merging**